### ISSUE TITLE
Migration from 'old' alerts to unified alerting is awful

### ISSUE THREAD

**Issue by fsperling** [grafana/grafana#45443]
Hello,

we would like to use the [HA feature](https://grafana.com/docs/grafana/latest/alerting/unified-alerting/high-availability/) in order to run multiple Grafana pods and not have a downtime during upgrades.

When I tried out the new unified alerting I was happy to see that:
- notification channels were migrated to contact points
- and that the connection between alerts and contact points was preserved (as opposed to having all alerts go to a default contact point)

However the details of the migration render the alerting unusable afterwards:
- for every combination of notification channels a contact point is generated (maybe even for every alert a contact point is generated)
  - Alert A with notification channel A will get a contact point A
  - Alert B with notification channel A and B will get a contact point B
  - Alert C with notification channel B will get a contact point C
- Thus way to many contact points are generated
- All contact points are called: autogen-contact-point-{\d}
- To connect Alert rules with contact points there are generated notification policies
   - those have a rule with a random uid which is connected to a contact point with a autogen-contact-point-{\d} name

This is completely un-maintainable.
It would be great if you could improve the migration to unified alerting otherwise I don't think it is an option for existing installations to switch to it.

**Comment by armandgrillet** [comment]
That's great feedback, we had the same issue with each alert being in their own folder on migration in previous 8.x releases. Alerting engineers will work on this in the coming month, there should be a way to deduplicate contact points during migration.

**Comment by JacobsonMT** [comment]
Hi @fsperling, thanks for reporting the issue! I just put up a PR (https://github.com/grafana/grafana/pull/47291) that I think addresses most of your comments (screenshots included).

There are a couple of caveats though:

* I didn't tackle the matching label on notification policy using a random rule_uid. I agree that this isn't optimal, it's difficult to see at a glance which alert each policy is supposed to reference. That being said, I can't think of a simple solution that doesn't have major drawbacks. For example, I could instead match using the rule title but this value isn't guaranteed to be unique and furthermore if the alert title is changed it should (correctly but silently) break the association. IMO, this behaviour is not intuitive for a migrated alert (especially for someone migrated to unified alerting for the first time). If you have thoughts here, I'm all ears!

* I had two options when legacy alerts had multiple default channels setup:
	1. Completely deduplicate contact points but have the notification policy structure become significantly more complex by adding a third level of nesting to all routes. (eventually I think letting notification policies directly have multiple contact points is the way to go here)
	2. Add a redundant contact point containing all legacy default alerts that is only used in the root policy.

I opted to go with ii. and keep a single redundant contact point (autogen-contact-point-default) in this case.

Let me know if you think the points I addressed in the PR help here and if you have any other thoughts on the state of legacy -> unified alerting migration, I'd love to hear them.
